File: sdk/go/searei.go
// Package searei provides the official Go client for SeaRei AI Safety API
//
// Installation:
//
//	go get github.com/searei/searei-go
//
// Usage:
//
//	import "github.com/searei/searei-go"
//
//	client := searei.NewClient(&searei.Config{
//	    APIKey: "your_api_key",
//	})
//
//	result, err := client.Score(context.Background(), "I will kill you", nil)
//	if err != nil {
//	    log.Fatal(err)
//	}
//
//	if result.IsUnsafe() {
//	    fmt.Printf("Flagged: %s (%.2f)\n", result.Category, result.Score)
//	}
package searei

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"time"
)

const (
	// Version is the current SDK version
	Version = "0.1.0"

	// DefaultBaseURL is the default API base URL
	DefaultBaseURL = "http://localhost:8001"

	// DefaultTimeout is the default request timeout
	DefaultTimeout = 30 * time.Second

	// DefaultMaxRetries is the default maximum number of retries
	DefaultMaxRetries = 3
)

// Config holds the client configuration
type Config struct {
	// APIKey for authentication (optional for self-hosted)
	APIKey string

	// BaseURL for the API (default: http://localhost:8001)
	BaseURL string

	// HTTPClient to use for requests (optional)
	HTTPClient *http.Client

	// MaxRetries for failed requests (default: 3)
	MaxRetries int
}

// ScoreResult represents the result of a content safety score request
type ScoreResult struct {
	Prediction      string  `json:"prediction"`
	Score           float64 `json:"score"`
	Category        string  `json:"category,omitempty"`
	Rationale       string  `json:"rationale,omitempty"`
	Method          string  `json:"method,omitempty"`
	LatencyMs       int     `json:"latency_ms,omitempty"`
	PolicyVersion   string  `json:"policy_version,omitempty"`
	PolicyChecksum  string  `json:"policy_checksum,omitempty"`
}

// IsSafe returns true if the content is safe (prediction == "pass")
func (r *ScoreResult) IsSafe() bool {
	return r.Prediction == "pass"
}

// IsUnsafe returns true if the content is unsafe (prediction == "flag")
func (r *ScoreResult) IsUnsafe() bool {
	return r.Prediction == "flag"
}

// String returns a string representation of the result
func (r *ScoreResult) String() string {
	return fmt.Sprintf("ScoreResult(prediction='%s', score=%.3f, category='%s')",
		r.Prediction, r.Score, r.Category)
}

// BatchResult represents the result of a batch score request
type BatchResult struct {
	Results   []*ScoreResult `json:"results"`
	LatencyMs int            `json:"latency_ms,omitempty"`
}

// Total returns the total number of texts scored
func (b *BatchResult) Total() int {
	return len(b.Results)
}

// Flagged returns the number of flagged texts
func (b *BatchResult) Flagged() int {
	count := 0
	for _, r := range b.Results {
		if r.IsUnsafe() {
			count++
		}
	}
	return count
}

// Safe returns the number of safe texts
func (b *BatchResult) Safe() int {
	return b.Total() - b.Flagged()
}

// String returns a string representation of the result
func (b *BatchResult) String() string {
	return fmt.Sprintf("BatchResult(total=%d, flagged=%d, safe=%d)",
		b.Total(), b.Flagged(), b.Safe())
}

// ScoreOptions holds options for score requests
type ScoreOptions struct {
	// Category filter (violence, harassment, etc.)
	Category string `json:"category,omitempty"`

	// Language code (en, es, ar, etc.)
	Language string `json:"language,omitempty"`
}

// Client is the SeaRei API client
type Client struct {
	apiKey     string
	baseURL    string
	httpClient *http.Client
	maxRetries int
}

// NewClient creates a new SeaRei API client
func NewClient(config *Config) *Client {
	if config == nil {
		config = &Config{}
	}

	apiKey := config.APIKey
	if apiKey == "" {
		apiKey = os.Getenv("SEAREI_API_KEY")
	}

	baseURL := config.BaseURL
	if baseURL == "" {
		baseURL = DefaultBaseURL
	}

	httpClient := config.HTTPClient
	if httpClient == nil {
		httpClient = &http.Client{
			Timeout: DefaultTimeout,
		}
	}

	maxRetries := config.MaxRetries
	if maxRetries == 0 {
		maxRetries = DefaultMaxRetries
	}

	return &Client{
		apiKey:     apiKey,
		baseURL:    baseURL,
		httpClient: httpClient,
		maxRetries: maxRetries,
	}
}

// request makes an HTTP request with retry logic
func (c *Client) request(ctx context.Context, method, endpoint string, body interface{}) ([]byte, error) {
	url := c.baseURL + endpoint

	var reqBody io.Reader
	if body != nil {
		jsonData, err := json.Marshal(body)
		if err != nil {
			return nil, fmt.Errorf("failed to marshal request body: %w", err)
		}
		reqBody = bytes.NewBuffer(jsonData)
	}

	var lastErr error
	for attempt := 0; attempt < c.maxRetries; attempt++ {
		req, err := http.NewRequestWithContext(ctx, method, url, reqBody)
		if err != nil {
			return nil, fmt.Errorf("failed to create request: %w", err)
		}

		req.Header.Set("User-Agent", "searei-go/"+Version)
		req.Header.Set("Content-Type", "application/json")

		if c.apiKey != "" {
			req.Header.Set("Authorization", "Bearer "+c.apiKey)
		}

		resp, err := c.httpClient.Do(req)
		if err != nil {
			lastErr = err
			if attempt < c.maxRetries-1 {
				time.Sleep(time.Duration(1<<uint(attempt)) * time.Second) // Exponential backoff
				continue
			}
			return nil, fmt.Errorf("request failed after %d retries: %w", c.maxRetries, err)
		}
		defer resp.Body.Close()

		respBody, err := io.ReadAll(resp.Body)
		if err != nil {
			return nil, fmt.Errorf("failed to read response body: %w", err)
		}

		// Handle errors
		if resp.StatusCode == http.StatusUnauthorized {
			return nil, &AuthenticationError{Message: "Invalid or missing API key"}
		} else if resp.StatusCode == http.StatusTooManyRequests {
			return nil, &RateLimitError{Message: "Rate limit exceeded"}
		} else if resp.StatusCode == http.StatusUnprocessableEntity {
			return nil, &ValidationError{Message: fmt.Sprintf("Validation error: %s", string(respBody))}
		} else if resp.StatusCode >= 400 {
			return nil, &APIError{StatusCode: resp.StatusCode, Message: string(respBody)}
		}

		return respBody, nil
	}

	return nil, fmt.Errorf("request failed after %d retries: %w", c.maxRetries, lastErr)
}

// Score scores a single text for safety violations
//
// Example:
//
//	result, err := client.Score(ctx, "I will kill you", &searei.ScoreOptions{
//	    Category: "violence",
//	})
func (c *Client) Score(ctx context.Context, text string, options *ScoreOptions) (*ScoreResult, error) {
	payload := map[string]interface{}{
		"text": text,
	}

	if options != nil {
		if options.Category != "" {
			payload["category"] = options.Category
		}
		if options.Language != "" {
			payload["language"] = options.Language
		}
	}

	respBody, err := c.request(ctx, http.MethodPost, "/score", payload)
	if err != nil {
		return nil, err
	}

	var result ScoreResult
	if err := json.Unmarshal(respBody, &result); err != nil {
		return nil, fmt.Errorf("failed to unmarshal response: %w", err)
	}

	return &result, nil
}

// BatchScore scores multiple texts in a single request
//
// Example:
//
//	texts := []string{"Hello world", "I will kill you", "Nice weather"}
//	result, err := client.BatchScore(ctx, texts, nil)
func (c *Client) BatchScore(ctx context.Context, texts []string, options *ScoreOptions) (*BatchResult, error) {
	rows := make([]map[string]string, len(texts))
	for i, text := range texts {
		rows[i] = map[string]string{"text": text}
	}

	payload := map[string]interface{}{
		"rows": rows,
	}

	if options != nil {
		if options.Category != "" {
			payload["category"] = options.Category
		}
		if options.Language != "" {
			payload["language"] = options.Language
		}
	}

	respBody, err := c.request(ctx, http.MethodPost, "/score-batch", payload)
	if err != nil {
		return nil, err
	}

	var result BatchResult
	if err := json.Unmarshal(respBody, &result); err != nil {
		return nil, fmt.Errorf("failed to unmarshal response: %w", err)
	}

	return &result, nil
}

// Health checks the API health status
//
// Example:
//
//	health, err := client.Health(ctx)
//	fmt.Println(health["status"])
func (c *Client) Health(ctx context.Context) (map[string]interface{}, error) {
	respBody, err := c.request(ctx, http.MethodGet, "/healthz", nil)
	if err != nil {
		return nil, err
	}

	var health map[string]interface{}
	if err := json.Unmarshal(respBody, &health); err != nil {
		return nil, fmt.Errorf("failed to unmarshal response: %w", err)
	}

	return health, nil
}

// Error types

// APIError represents a generic API error
type APIError struct {
	StatusCode int
	Message    string
}

func (e *APIError) Error() string {
	return fmt.Sprintf("API error (%d): %s", e.StatusCode, e.Message)
}

// AuthenticationError represents an authentication error
type AuthenticationError struct {
	Message string
}

func (e *AuthenticationError) Error() string {
	return e.Message
}

// RateLimitError represents a rate limit error
type RateLimitError struct {
	Message string
}

func (e *RateLimitError) Error() string {
	return e.Message
}

// ValidationError represents a validation error
type ValidationError struct {
	Message string
}

func (e *ValidationError) Error() string {
	return e.Message
}

// Quick score function without creating a client
func Score(ctx context.Context, text string, config *Config, options *ScoreOptions) (*ScoreResult, error) {
	client := NewClient(config)
	return client.Score(ctx, text, options)
}

File: sdk/python/searei.py
#!/usr/bin/env python3
"""
SeaRei Python SDK
Official Python client for SeaRei AI Safety API

Installation:
    pip install searei

Usage:
    from searei import SeaReiClient
    
    client = SeaReiClient(api_key="your_api_key")
    result = client.score("I will kill you")
    
    if result.is_unsafe:
        print(f"Flagged: {result.category} ({result.score:.2f})")
"""

import os
import time
from typing import Any, Dict, List, Optional, Union
from dataclasses import dataclass, field

try:
    import requests
except ImportError:
    raise ImportError("requests library required. Install with: pip install requests")


__version__ = "0.1.0"
__author__ = "SeaRei Team"


@dataclass
class ScoreResult:
    """Result from a content safety score request"""
    prediction: str
    score: float
    category: Optional[str] = None
    rationale: Optional[str] = None
    method: Optional[str] = None
    latency_ms: Optional[int] = None
    policy_version: Optional[str] = None
    policy_checksum: Optional[str] = None
    
    @property
    def is_safe(self) -> bool:
        """Returns True if content is safe (prediction == 'pass')"""
        return self.prediction == "pass"
    
    @property
    def is_unsafe(self) -> bool:
        """Returns True if content is unsafe (prediction == 'flag')"""
        return self.prediction == "flag"
    
    def __repr__(self) -> str:
        return f"ScoreResult(prediction='{self.prediction}', score={self.score:.3f}, category='{self.category}')"


@dataclass
class BatchResult:
    """Result from a batch score request"""
    results: List[ScoreResult]
    total: int
    flagged: int
    safe: int
    latency_ms: int
    
    def __repr__(self) -> str:
        return f"BatchResult(total={self.total}, flagged={self.flagged}, safe={self.safe})"


class SeaReiError(Exception):
    """Base exception for SeaRei SDK errors"""
    pass


class AuthenticationError(SeaReiError):
    """Raised when API key is invalid or missing"""
    pass


class RateLimitError(SeaReiError):
    """Raised when rate limit is exceeded"""
    pass


class ValidationError(SeaReiError):
    """Raised when request validation fails"""
    pass


class SeaReiClient:
    """
    SeaRei AI Safety API Client
    
    Args:
        api_key: API key for authentication (optional for self-hosted)
        base_url: Base URL for API (default: http://localhost:8001)
        timeout: Request timeout in seconds (default: 30)
        max_retries: Maximum number of retries on failure (default: 3)
    
    Example:
        >>> client = SeaReiClient(api_key="sk_test_...")
        >>> result = client.score("I will kill you")
        >>> print(result.prediction)
        flag
    """
    
    def __init__(
        self,
        api_key: Optional[str] = None,
        base_url: str = "http://localhost:8001",
        timeout: int = 30,
        max_retries: int = 3
    ):
        self.api_key = api_key or os.environ.get("SEAREI_API_KEY")
        self.base_url = base_url.rstrip("/")
        self.timeout = timeout
        self.max_retries = max_retries
        self._session = requests.Session()
        
        # Set default headers
        self._session.headers.update({
            "User-Agent": f"searei-python/{__version__}",
            "Content-Type": "application/json"
        })
        
        if self.api_key:
            self._session.headers["Authorization"] = f"Bearer {self.api_key}"
    
    def _request(self, method: str, endpoint: str, **kwargs) -> Dict[str, Any]:
        """Make an HTTP request with retry logic"""
        url = f"{self.base_url}{endpoint}"
        kwargs.setdefault("timeout", self.timeout)
        
        last_error = None
        for attempt in range(self.max_retries):
            try:
                response = self._session.request(method, url, **kwargs)
                
                # Handle errors
                if response.status_code == 401:
                    raise AuthenticationError("Invalid or missing API key")
                elif response.status_code == 429:
                    raise RateLimitError("Rate limit exceeded")
                elif response.status_code == 422:
                    raise ValidationError(f"Validation error: {response.text}")
                elif response.status_code >= 400:
                    raise SeaReiError(f"API error ({response.status_code}): {response.text}")
                
                return response.json()
                
            except requests.exceptions.RequestException as e:
                last_error = e
                if attempt < self.max_retries - 1:
                    time.sleep(2 ** attempt)  # Exponential backoff
                continue
        
        raise SeaReiError(f"Request failed after {self.max_retries} retries: {last_error}")
    
    def score(
        self,
        text: str,
        category: Optional[str] = None,
        language: Optional[str] = None
    ) -> ScoreResult:
        """
        Score a single text for safety violations
        
        Args:
            text: Text content to score
            category: Optional category filter (violence, self_harm, harassment, etc.)
            language: Optional language code (en, es, ar, etc.)
        
        Returns:
            ScoreResult with prediction, score, and metadata
        
        Example:
            >>> result = client.score("I will kill you")
            >>> if result.is_unsafe:
            ...     print(f"Flagged: {result.category}")
        """
        payload = {"text": text}
        if category:
            payload["category"] = category
        if language:
            payload["language"] = language
        
        data = self._request("POST", "/score", json=payload)
        
        return ScoreResult(
            prediction=data["prediction"],
            score=data["score"],
            category=data.get("category"),
            rationale=data.get("rationale"),
            method=data.get("method"),
            latency_ms=data.get("latency_ms"),
            policy_version=data.get("policy_version"),
            policy_checksum=data.get("policy_checksum")
        )
    
    def batch_score(
        self,
        texts: List[str],
        category: Optional[str] = None,
        language: Optional[str] = None
    ) -> BatchResult:
        """
        Score multiple texts in a single request
        
        Args:
            texts: List of text contents to score
            category: Optional category filter
            language: Optional language code
        
        Returns:
            BatchResult with results for all texts
        
        Example:
            >>> texts = ["Hello world", "I will kill you", "Nice weather"]
            >>> result = client.batch_score(texts)
            >>> print(f"Flagged {result.flagged}/{result.total}")
        """
        rows = [{"text": text} for text in texts]
        payload = {"rows": rows}
        if category:
            payload["category"] = category
        if language:
            payload["language"] = language
        
        data = self._request("POST", "/score-batch", json=payload)
        
        results = [
            ScoreResult(
                prediction=r["prediction"],
                score=r["score"],
                category=r.get("category"),
                rationale=r.get("rationale"),
                method=r.get("method")
            )
            for r in data["results"]
        ]
        
        return BatchResult(
            results=results,
            total=len(results),
            flagged=sum(1 for r in results if r.is_unsafe),
            safe=sum(1 for r in results if r.is_safe),
            latency_ms=data.get("latency_ms", 0)
        )
    
    def health(self) -> Dict[str, Any]:
        """
        Check API health status
        
        Returns:
            Dict with health status and metadata
        
        Example:
            >>> health = client.health()
            >>> print(health["status"])
            healthy
        """
        return self._request("GET", "/healthz")
    
    def __enter__(self):
        """Context manager support"""
        return self
    
    def __exit__(self, exc_type, exc_val, exc_tb):
        """Close session on context exit"""
        self._session.close()


# Convenience function
def score(text: str, api_key: Optional[str] = None, **kwargs) -> ScoreResult:
    """
    Quick score function without creating a client
    
    Example:
        >>> from searei import score
        >>> result = score("I will kill you")
        >>> print(result.prediction)
        flag
    """
    with SeaReiClient(api_key=api_key) as client:
        return client.score(text, **kwargs)


if __name__ == "__main__":
    # Simple CLI for testing
    import sys
    
    if len(sys.argv) < 2:
        print("Usage: python searei.py 'text to score'")
        sys.exit(1)
    
    text = " ".join(sys.argv[1:])
    
    try:
        result = score(text)
        print(f"\nResult: {result.prediction}")
        print(f"Score: {result.score:.3f}")
        if result.category:
            print(f"Category: {result.category}")
        if result.rationale:
            print(f"Rationale: {result.rationale}")
        print(f"Method: {result.method}")
        print(f"Latency: {result.latency_ms}ms")
    except SeaReiError as e:
        print(f"Error: {e}", file=sys.stderr)
        sys.exit(1)












